test/e2e/apimachinery/watch.go

```diff
@@ -33,158 +31,258 @@ import (
 		Expect(err).NotTo(HaveOccurred())
-		expectEvent(watchA, watch.Added, testPodA)
-		expectEvent(watchAB, watch.Added, testPodA)
-		expectNoEvent(watchB, watch.Added, testPodA)
+		expectEvent(watchA, watch.Added, testConfigMapA)
+		expectEvent(watchAB, watch.Added, testConfigMapA)
+		expectNoEvent(watchB, watch.Added, testConfigMapA)
 
-		testPodA, err = updatePod(f, testPodA.GetName(), func(p *v1.Pod) {
-			p.ObjectMeta.Labels["mutation"] = "1"
+		By("modifying configmap A and ensuring the correct watchers observe the notification")
+		testConfigMapA, err = updateConfigMap(c, ns, testConfigMapA.GetName(), func(cm *v1.ConfigMap) {
+			setConfigMapData(cm, "mutation", "1")
 		})
 		Expect(err).NotTo(HaveOccurred())
-		expectEvent(watchA, watch.Modified, testPodA)
-		expectEvent(watchAB, watch.Modified, testPodA)
-		expectNoEvent(watchB, watch.Modified, testPodA)
+		expectEvent(watchA, watch.Modified, testConfigMapA)
+		expectEvent(watchAB, watch.Modified, testConfigMapA)
+		expectNoEvent(watchB, watch.Modified, testConfigMapA)
 
-		testPodA, err = updatePod(f, testPodA.GetName(), func(p *v1.Pod) {
-			p.ObjectMeta.Labels["mutation"] = "2"
+		By("modifying configmap A again and ensuring the correct watchers observe the notification")
+		testConfigMapA, err = updateConfigMap(c, ns, testConfigMapA.GetName(), func(cm *v1.ConfigMap) {
+			setConfigMapData(cm, "mutation", "2")
 		})
 		Expect(err).NotTo(HaveOccurred())
-		expectEvent(watchA, watch.Modified, testPodA)
-		expectEvent(watchAB, watch.Modified, testPodA)
-		expectNoEvent(watchB, watch.Modified, testPodA)
+		expectEvent(watchA, watch.Modified, testConfigMapA)
+		expectEvent(watchAB, watch.Modified, testConfigMapA)
+		expectNoEvent(watchB, watch.Modified, testConfigMapA)
 
-		err = c.CoreV1().Pods(ns).Delete(testPodA.GetName(), nil)
+		By("deleting configmap A and ensuring the correct watchers observe the notification")
+		err = c.CoreV1().ConfigMaps(ns).Delete(testConfigMapA.GetName(), nil)
 		Expect(err).NotTo(HaveOccurred())
 		expectEvent(watchA, watch.Deleted, nil)
 		expectEvent(watchAB, watch.Deleted, nil)
 		expectNoEvent(watchB, watch.Deleted, nil)
 
-		testPodB, err = c.CoreV1().Pods(ns).Create(testPodB)
+		By("creating a configmap with label B and ensuring the correct watchers observe the notification")
+		testConfigMapB, err = c.CoreV1().ConfigMaps(ns).Create(testConfigMapB)
 		Expect(err).NotTo(HaveOccurred())
-		expectEvent(watchB, watch.Added, testPodB)
-		expectEvent(watchAB, watch.Added, testPodB)
-		expectNoEvent(watchA, watch.Added, testPodB)
+		expectEvent(watchB, watch.Added, testConfigMapB)
+		expectEvent(watchAB, watch.Added, testConfigMapB)
+		expectNoEvent(watchA, watch.Added, testConfigMapB)
 
-		err = c.CoreV1().Pods(ns).Delete(testPodB.GetName(), nil)
+		By("deleting configmap B and ensuring the correct watchers observe the notification")
+		err = c.CoreV1().ConfigMaps(ns).Delete(testConfigMapB.GetName(), nil)
```

Comment:
One more thing that I would like to see tested (sorry that I didn't mention it on the first time, but I realized in the meantime that it's an extremely important case and we used to have bugs about that in the past is):

- you create an object with label - add watch event should be send
- you delete that label (or change its value) - delete watch event should be send
- you add the label back - add event should be send this time
- you delete the object - delete watch event should be send.

This should probably be a separate (fourth) test.

Reply:
also done